File: POO/TP/POO_TP3/TP3/justeprix.py
import random

from TP3 import utils, menu
import time

# Une partie est contenue dans une classe pour plus de simplicité et de lisibilité au niveau des variables.
class Game:
    # `minBound` et `maxBound` représente les valeurs minimale et maximale du nombre généré.
    # `maxAttempts` représente le nombre maximum d'essais pour l'utilisateur.
    def __init__(self, minBound:int=-100, maxBound:int=100, maxAttempts:int=15):
        self.minBound = minBound
        self.maxBound = maxBound
        self.number = random.randint(minBound, maxBound)
        self.maxAttempts = maxAttempts
        self.userAttempts = 1

    # Lance le jeu lui-même.
    # `debug`=True permet d'afficher le nombre mystère dès le début de la partie.
    def launchRounds(self, debug:bool=False):
        if debug is True:
            print(f"DEBUG MODE ON || Nombre généré: {self.number}")

        # Représente les rounds de jeu.
        while self.userAttempts <= self.maxAttempts:
            print(f"ESSAIS RESTANTS: {self.maxAttempts - self.userAttempts + 1}")

            userGuess = utils.getUserInput(self.minBound, self.maxBound, f"Quel est le nombre mystère? [{self.minBound}:{self.maxBound}] → ")

            # Condition de victoire du jeu (stoppe l'exécution de la fonction).
            if userGuess == self.number:
                self.isWon()
                return

            # Evaluation du nombre si le jeu n'a pas été gagné.
            elif userGuess > self.number:
                print("Le nombre est plus petit !")
            else:
                print("Le nombre est plus grand !")

            utils.jumpLines()
            self.userAttempts += 1

        # Si le programme arrive à ce point de la fonction, c'est que le joueur a perdu.
        # Condition de défaite.
        self.isLost()

    # Print le résultat du jeu. `waitingTime` permet un temps de pause avant la suite du programme.
    def isWon(self, waitingTime:float=1.3):
        print(f"Félicitations !! C'est une victoire en {self.userAttempts} essai(s) !")
        utils.jumpLines()
        time.sleep(waitingTime)
    def isLost(self, waitingTime:float=1.3):
        print(f"Raté... Le nombre mystère était {self.number}.")
        utils.jumpLines()
        time.sleep(waitingTime)

def main():
    userRetry = -1
    while userRetry != 0:
        menu.Box("L E  J U S T E  P R I X").drawBox()
        utils.jumpLines()

        # Fonction du jeu.
        Game().launchRounds()

        # Menu retry (0 = quitter, 1 = réessayer).
        userRetry = menu.drawBinaryMenu("Lancer une nouvelle partie ?")
        utils.jumpLines()

File: POO/TP/POO_TP2/src/JustePrix.java
public class JustePrix {

    // Génère un nombre aléatoire situé entre `min` et `max` tous deux inclus.
    public static int generateRandomInt(int min, int max) {
        return (int)Math.floor(Math.random() * (max - min + 1) + min);
    }

    // Gère une partie de Juste Prix.
    // `maxAttempts` définit le nombre d'essais maximum pour le joueur.
    // `minBound` et `maxBound` définissent les limites du chiffre à deviner.
    // Plus la différence entre ces deux valeurs est grande et le nombre d'essai petit,
    // plus le jeu sera difficile.
    // `debug` est une variable permettant de print le nombre dès le départ de la partie.
    public static void playGame(int maxAttempts, int minBound, int maxBound, boolean debug) {
        Utils.drawTitleBox(48, "L E  J U S T E  P R I X", "POO | TP n°1", 0);
        // Génère le nombre à deviner.
        int number = generateRandomInt(-100, 100);
        if (debug) {
            System.out.printf("DEBUG MODE ON || Nombre généré: %s\n\n", number);
        }
        // userAttempts est la valeur "contraire" à maxAttempts.
        // Deux variables sont utilisées pour traquer le nombre d'essais à la fin de la partie.
        int userAttempts = 1;
        int userGuess;

        // Cette boucle s'exécute tant que le joueur peut encore jouer.
        // Si le joueur gagne, la fonction s'arrête immédiatement.
        for (; userAttempts <= maxAttempts; userAttempts++) {
            System.out.printf("\nATTEMPTS LEFT: %d\n", maxAttempts +1 - userAttempts);
            userGuess = Utils.askForInt(minBound, maxBound, "Quel est le nombre mystère ? → ", false);

            // Victoire: le joueur a trouvé le nombre mystère.
            if (userGuess == number) {
                System.out.printf("\nFélicitations !! C'est une victoire en %d essai(s) !\n", userAttempts);
                Utils.sleepTime(1200);
                return;
            // Round perdu: chiffre deviné trop grand ou trop petit.
            } else if (userGuess > number) {
                System.out.println("C'est plus petit !");
            } else {
                System.out.println("C'est plus grand !");
            }
        }

        // Défaite: le joueur n'a pas trouvé le nombre dans son nombre d'essais impartis.
        System.out.printf("\nRaté... Le nombre mystère était %d.\n", number);
        Utils.sleepTime(1200);
    }

    public static void main(String[] args) {
        // Pour une UX plus fluide, le titleBox se trouve en haut de la fonction `playGame` et non ici.
        System.out.println();

        // Définit la valeur par défaut pour le menu Retry.
        // Les valeurs ne peuvent être que 0 et 1, -1 initialise la variable pour que la boucle while
        // se lance, avec une valeur qui ne peut pas bloquer son exécution.
        int retry = -1;
        while (retry != 0) {
            // Lance le jeu.
            // `debug` permet de print le nombre généré en haut du jeu.
            playGame(10, -100, 100, false);

            // Retry: 1 relance la fonction depuis le début (boucle while), 0 l'arrête
            // et revient au menu précédent.
            System.out.println("Lancer une nouvelle partie ?\n");
            Menu.runMenu(true, null);
            retry = Utils.askForInt(0, 1, "→ ", false);
            System.out.println();
        }
    }
}
